### PR TITLE
New Services List Page & Create/Edit Screen 

### PR DESCRIPTION
**Pull request by rgallettonf**
* List page with ag-grid for displaying services
* Create/Edit screen for defining services
* Generation of inputs/form fields when adding Configs to a service based on selected Config Type schema
* Toggle between Form View and JSON View when editing a config

Screen Shot 2024-01-16 at 7 56 58 AM

### CODE REVIEW

File: server-edge.js [hunk 1]
```diff
@@ -44,7 +44,8 @@ var helmetOptions = {
       },
     },
     frameguard: { action: 'SAMEORIGIN' },
-    crossOriginEmbedderPolicy: false
+    crossOriginEmbedderPolicy: false,
+    strictTransportSecurity: false
```
Comment (dovholuknf) [inline server-edge.js:48]: this change doesn't seem related to the PR. Is it intentional? Does this "help/fix" the http/https problems we've had?

Comment (rgallettonf) [reply server-edge.js:48]: Nope good catch, this must have snuck in when I was troubleshooting that issue. I will revert

File: projects/ziti-console-lib/src/lib/pages/services/services-page.service.ts [hunk 2]
```diff
@@ -0,0 +1,198 @@
+import {Injectable, Inject} from "@angular/core";
```
Comment (dovholuknf) [inline projects/ziti-console-lib/src/lib/pages/services/services-page.service.ts:1]: missing copyright

File: projects/ziti-console-lib/src/lib/pages/services/services-page.component.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
+import {Component, Inject, OnInit, OnDestroy} from '@angular/core';
```
Comment (dovholuknf) [inline projects/ziti-console-lib/src/lib/pages/services/services-page.component.ts:1]: i won't make this comment again- missing copyright. please check all net-new files for that header.

File: projects/ziti-console-lib/src/lib/pages/services/services-page.service.ts [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        });
+    }
+
+    public getServiceRoleAttributes() {
+        return this.zitiService.get('service-role-attributes', {}, []);
+    }
+
+    downloadAllItems() {
+        const paging = cloneDeep(this.paging);
+        paging.total = 2000;
```
Comment (dovholuknf) [inline projects/ziti-console-lib/src/lib/pages/services/services-page.service.ts:162]: what's up with the hardcoded "2000"?

File: projects/ziti-console-lib/src/lib/pages/services/services-page.component.ts (hunk 3 above)
Comment (rgallettonf) [reply projects/ziti-console-lib/src/lib/pages/services/services-page.component.ts:1]: Will do

File: projects/ziti-console-lib/src/lib/pages/services/services-page.service.ts [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        const rolesRenderer = (row) => {
+            let roles = '';
+            row?.data?.roleAttributes?.forEach((attr) => {
+                roles += '<div class="hashtag">'+attr+'</div>';
+            });
+            return roles;
+        }
+
+        const createdAtFormatter = (row) => {
+            return moment(row?.data?.createdAt).local().format('M/D/YYYY H:MM A');
```
Comment (dovholuknf) [inline projects/ziti-console-lib/src/lib/pages/services/services-page.service.ts:82]: can we set this up for internationalization? this is most definitely american-centric...

File: projects/ziti-console-lib/src/lib/pages/services/services-page.component.scss [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+      display: flex;
+      flex-direction: row;
+      align-items: center;
+    }
+
+    .v7Offline {
+      background-color: var(--placeholder);
+    }
+
+    .v7OnlineOffline {
```
Comment (dovholuknf) [inline projects/ziti-console-lib/src/lib/pages/services/services-page.component.scss:214]: ".v7"? does this sorta stuff belong here?

File: projects/ziti-console-lib/src/lib/features/projectable-forms/service/service-form.component.ts [hunk 7]
```diff
@@ -0,0 +1,561 @@
...
+  _dataChange = false;
+  watchData() {
+    delay(() => {
+      const dataChange = !isEqual(this.initData, this.formData);
+      if (dataChange !== this._dataChange) {
+        this.dataChange.emit(dataChange);
+      }
+      this._dataChange = dataChange;
+      this.watchData();
+    }, 100);
```
Comment (dovholuknf) [inline projects/ziti-console-lib/src/lib/features/projectable-forms/service/service-form.component.ts:559]: this seems like the sort of thing that should be a component or at least a function abstracted away, no? also this is polling? :/ Is it making a request every 100 ms?

File: projects/ziti-console-lib/src/lib/features/projectable-forms/service/service-form.component.ts [hunk 8]
```diff
@@ -0,0 +1,561 @@
...
+      this.configErrors['name'] = true;
+    }
+    return isEmpty(this.configErrors);
+  }
+
+  toggleEncryptionRequired() {
+    this.formData.encryptionRequired = !this.formData.encryptionRequired;
+  }
+
+  closeModal(refresh = true, ignoreChanges = false): void {
```
Comment (dovholuknf) [inline projects/ziti-console-lib/src/lib/features/projectable-forms/service/service-form.component.ts:537]: seems like the kinda function that should be reused?

File: projects/ziti-console-lib/src/lib/features/dynamic-widgets/port-ranges/port-ranges.component.ts [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  getProperties() {
+    if (!this.fieldValue) {
+      return [];
+    }
+    const ranges = [];
+    this.fieldValue.forEach((val: string) => {
+      const vals = val.split('-');
+      if (vals.length === 1) {
```
Comment (dovholuknf) [inline projects/ziti-console-lib/src/lib/features/dynamic-widgets/port-ranges/port-ranges.component.ts:34]: can i assume that somewhere else, something has validated that 'val' can only be numbers and that those numbers are only from 1 to 65535 ? even still, i'd say this is the kind of code that should be separated on its own and reused as needed.

File: projects/ziti-console-lib/src/lib/features/dynamic-widgets/port-ranges/port-ranges.component.scss
Comment (dovholuknf) [inline projects/ziti-console-lib/src/lib/features/dynamic-widgets/port-ranges/port-ranges.component.scss:1]: empty file? i assume it's just 'required'?

**Comment by dovholuknf** [comment]
definitely some things to fix before merging

**Review by dovholuknf — APPROVED** [review]

File: projects/ziti-console-lib/src/lib/features/dynamic-widgets/port-ranges/port-ranges.component.scss
Comment (rgallettonf) [reply projects/ziti-console-lib/src/lib/features/dynamic-widgets/port-ranges/port-ranges.component.scss:1]: Yep, this is a generated file when creating a new component. No harm in keeping it for now so it's readily available if/when we need to make styling adjustments to the componet.

File: projects/ziti-console-lib/src/lib/features/dynamic-widgets/forwarding-config/forwarding-config.service.ts [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+        ];
+        return props;
+    }
+
+    getPortRanges(allowedPortRanges) {
+        if (!allowedPortRanges) {
+            return [];
+        }
+        const ranges = [];
+        allowedPortRanges.forEach((val: string) => {
```
[the remark this answers is not in the document]
Comment (rgallettonf) [reply projects/ziti-console-lib/src/lib/features/dynamic-widgets/forwarding-config/forwarding-config.service.ts:44]: Agreed, some of this logic around port-ranges can be consolidated into a shared service.

File: projects/ziti-console-lib/src/lib/features/dynamic-widgets/protocol-address-port/protocol-address-port-input.component.ts [hunk 11]
```diff
@@ -1,55 +1,81 @@
...
-      this.fieldValueChange.emit(data);
-      this.valueChange.next(data);
-    }, 500)();
+  ngOnInit() {
+    const getVals = () => {
+      let val = this.address;
+      this.addressChange.emit(this.address);
+      setTimeout(() => {
+        getVals();
+      }, 2000);
```
[the remark this answers is not in the document]
Comment (rgallettonf) [reply projects/ziti-console-lib/src/lib/features/dynamic-widgets/protocol-address-port/protocol-address-port-input.component.ts:47]: I will revisit this. May have been introduced during some debugging. Anything with `setTimeout()` has a bit of a code-smell to it for sure.

**Comment by rgallettonf** [comment]
I will be re-creating this PR to address the above comments as well as to apply proper commit signing to some of the changes that I made before I had signed commits setup on my local repo.
